randovania/games/binary_data.py

```diff
@@ -228,93 +222,112 @@ def _build_resource_info(**kwargs):
     z=Float64b,
 )
 
-ConstructNode = Struct(
-    name=CString("utf8"),
-    heal=Flag,
-    coordinates=OptionalValue(ConstructNodeCoordinates),
+NodeBaseFields = {
+    "heal": Flag,
+    "coordinates": OptionalValue(ConstructNodeCoordinates),
+    "extra": JsonEncodedValue,
+    "connections": ConstructDict(ConstructRequirement),
+}
+
+
+class NodeAdapter(construct.Adapter):
+    def _decode(self, obj: construct.Container, context, path):
+        result = construct.Container(node_type=obj["node_type"])
+        result.update(obj["data"])
+        result.move_to_end("connections")
+        return result
+
+    def _encode(self, obj: construct.Container, context, path):
+        data = copy.copy(obj)
+        return construct.Container(
+            node_type=data.pop("node_type"),
+            data=data,
+        )
+
+
+ConstructNode = NodeAdapter(Struct(
     node_type=construct.Enum(Byte, generic=0, dock=1, pickup=2, teleporter=3, event=4, translator_gate=5,
                              logbook=6, player_ship=7),
     data=Switch(
         lambda this: this.node_type,
         {
+            "generic": Struct(
+                **NodeBaseFields,
+            ),
             "dock": Struct(
+                **NodeBaseFields,
                 dock_index=Byte,
-                connected_area_asset_id=VarInt,
+                connected_area_name=String,
                 connected_dock_index=Byte,
                 dock_type=Byte,
                 dock_weakness_index=VarInt,
             ),
             "pickup": Struct(
+                **NodeBaseFields,
                 pickup_index=VarInt,
                 major_location=Flag,
             ),
             "teleporter": Struct(
-                destination_world_asset_id=VarInt,
-                destination_area_asset_id=VarInt,
-                teleporter_instance_id=OptionalValue(VarInt),
-                scan_asset_id=OptionalValue(VarInt),
+                **NodeBaseFields,
+                destination=ConstructAreaIdentifier,
                 keep_name_when_vanilla=Flag,
                 editable=Flag,
             ),
             "event": Struct(
+                **NodeBaseFields,
                 event_index=VarInt,
             ),
             "translator_gate": Struct(
+                **NodeBaseFields,
                 gate_index=VarInt,
             ),
             "logbook": Struct(
+                **NodeBaseFields,
                 string_asset_id=VarInt,
                 lore_type=ConstructLoreType,
-                extra=VarInt,
+                lore_extra=VarInt,
```

Comment:
`lore_extra` and `string_asset_id` should be moved to the `extra` field as well imo